File: ts/lang/otterop/lang/PureIterator.ts
import { OOPIterator } from "./OOPIterator";

export class PureIterator<T> implements Iterator<T> {
    private oopIterator : OOPIterator<T>;

    private constructor(oopIterator: OOPIterator<T>) {
        this.oopIterator = oopIterator;
    }

    next() : IteratorResult<T, number | undefined> {
        if (!this.oopIterator.hasNext()) {
            return {
                done: true,
                value: undefined
            }
        } else {
            return {
                done: false,
                value: this.oopIterator.next()
            }
        }
    }

    public static newIterator<T>(oopIterator: OOPIterator<T>) : Iterator<T> {
        return new PureIterator<T>(oopIterator);
    }
}

File: python/lang/otterop/lang/pure_iterator.py
class PureIterator:
    def __init__(self, oop_iterator):
        self.oop_iterator = oop_iterator

    def __next__(self):
        if not self.oop_iterator.has_next():
            raise StopIteration
        return self.oop_iterator.next();

    @staticmethod
    def new_iterator(oop_iterator):
        return PureIterator(oop_iterator)
